built_value/lib/iso_8601_duration_serializer.dart

```diff
@@ -0,0 +1,96 @@
+import 'package:built_collection/built_collection.dart';
+import 'package:built_value/serializer.dart';
+
+/// Alternative serializer for [Duration].
+///
+/// Install this to use ISO8601 compatible format instead of the default
+/// (microseconds). Use [SerializersBuilder.add] to install it.
+///
+/// Note that this serializer is not 100% compatible with the ISO8601 format
+/// due to limitations of the [Duration] class, but is designed to produce and
+/// consume reasonable strings that match the standard.
+class Iso8601DurationSerializer extends PrimitiveSerializer<Duration> {
+  @override
+  Duration deserialize(Serializers serializers, Object serialized,
+      {FullType specifiedType = FullType.unspecified}) {
+    final String value = serialized;
+    return _parseDuration(value);
+  }
+
+  @override
+  Object serialize(Serializers serializers, Duration object,
+      {FullType specifiedType = FullType.unspecified}) {
+    return _writeIso8601Duration(object);
+  }
+
+  @override
+  Iterable<Type> get types => BuiltList(const [Duration]);
+
+  @override
+  String get wireName => 'Duration';
+
+  Duration _parseDuration(String value) {
+    final match = _parseFormat.firstMatch(value);
+    if (match != null) {
+      // Iterate through the capture groups to build the unit mappings
+      final unitMappings = <String, int>{};
+
+      // Start iterating at 1, because match[0] is the full match
+      for (var i = 1; i <= match.groupCount; i++) {
+        final group = match[i];
+        if (group == null) {
+          continue;
+        }
+        // Get all but last character in group
+        // RegExp ensures this must be an int
+        final value = int.parse(group.substring(0, group.length - 1));
+        // Get last character
+        final unit = group.substring(group.length - 1);
+        unitMappings[unit] = value;
+      }
+      return Duration(
+        days: unitMappings[_dayToken] ?? 0,
+        hours: unitMappings[_hourToken] ?? 0,
+        minutes: unitMappings[_minuteToken] ?? 0,
+        seconds: unitMappings[_secondToken] ?? 0,
+      );
+    } else {
+      throw FormatException("Invalid duration format", value);
```

Comment:
Please move this up as a 'fail early' to reduce indentation.

if (match == null) throw FormatException(...).

Additionally, please use single quotes for strings.